explain how to create a machine (#28)
tweag/crem: docs/how-to-create-a-machine.md

# How to create a machine

A `StateMachine a b` is a stateful process which receives inputs of type `a` and emits outputs of type `b`.

We will represent a machine as such

```
a ┌────┐ b
──┤    ├──
  └────┘
```

as black boxes, where inputs are on the left and outputs are on the right.

The `StateMachine a b` data type has four constructors which we can use to construct a machine:

- `Basic`
- `Compose`
- `Parallel`
- `Alternative`

Let's start with the last three.

## `Compose`

```haskell
Compose
  :: StateMachine a b
  -> StateMachine b c
  -> StateMachine a c
```

allows us to sequentially compose two machines

```
a ┌────┐ b   b ┌────┐ c
──┤    ├──   ──┤    ├──
  └────┘       └────┘
```

to get a single machine

```
  ┌─────────────────┐
a │ ┌╌╌╌╌┐ b ┌╌╌╌╌┐ │ c
──┼╌┤    ├╌╌╌┤    ├╌┼──
  │ └╌╌╌╌┘   └╌╌╌╌┘ │
  └─────────────────┘

> ```

Reviewer: I certainly appreciate the effort that must have gone into drawing this intricate ASCII art. :laughing: 

Consider using [Mermaid](https://mermaid.js.org/syntax/stateDiagram.html) instead.

```mermaid
stateDiagram-v2
    direction LR
    [*] --> A
    state A {
        direction LR
        [*] --> B: a
        B --> C: b
        C --> [*]: c
    }
    A --> [*]
```

Author: I do love Mermaid, but for these diagram I kind of prefer to be able to control precisely how things are represented, especially for more complex diagrams like this one:

![Screenshot_20230131_095000](https://user-images.githubusercontent.com/2643972/215712893-cbd7eaef-6f5c-4c51-a18d-1c4f6cdbba28.png)

If in the future we will use Mermaid to generate similar diagrams automatically, then it would definitely make sense to use the same tool in the documentation

```
a ┌────┐ b
──┤    ├─>
  └────┘
```

as black boxes, where inputs are on the left and outputs are on the right.

The `StateMachine a b` data type has six constructors which we can use to construct a machine:

- `Basic`
- `Compose`
- `Parallel`
- `Alternative`
- `Feedback`
- `Kleisli`

> Let's start with the last five.

Reviewer: ```suggestion
Every state machine that is not `Basic` is a composition of two state machines, which are ultimately expressed in terms of `Basic`.

Let's skip over `Basic` for now, and look at the other constructors first.
```

This is why I think `Compose` should be named `Sequence`; all of these constructors are compositions, but not every composition is sequential.

Author: yes, now I agree with you. I'll do the change in a separate PR

In practice, if the composed machine receives a `Left a` as input, the `a` will be passed as input to the first machine that will emit a `b`, which will be wrapped to emit a `Left b` from the composed machine, while the second machine remains untouched.
Similarly, if the composed machine receives a `Right c` as input, the `c` will be passed as input to the second machine that will emit a `d`, which will be wrapped to emit a `Right d` from the composed machine, while the first machine remains untouched.

## `Feedback`

```haskell
Feedback
  :: StateMachine a [b]
  -> StateMachine b [a]
  -> StateMachine a [b]
```

allows us to use a machine to compute some new input to be processed from the output of a given machine.

If we have two machines

```
a ┌────┐ [b]
──┤    ├──>>
  └────┘
[a] ┌────┐ b
<<──┤    ├──
    └────┘
```

> We can compose them like so

Reviewer: Case in point that `Compose` is a misleading constructor name. :)

```

It works as follows. When an input `a` is received by the input machine, it is processed by the first machine to obtain a list of output `bs :: [b]`. These outputs as fed as inputs into the second machine one by one, obtaining a list of outputs `as :: [a]`. These outputs are then fed again to the first machine and the circle starts again, until one of the two machines emits the empty list `[]`. The composed machine will emit all the `bs` emitted in each round of the loop.

Notice that in fact everything works not only for lists, but for every foldable type.

## `Kleisli`

```haskell
Kleisli
  :: StateMachineT m a [b]
  -> StateMachineT m b [c]
  -> StateMachineT m a [c]
```

is very similar to `Compose`, but it allows us to compose machines which emit multiple outputs.

Consider two machines

```
a ┌────┐ [b]   b ┌────┐ [c]
──┤    ├──>>   ──┤    ├──>>
  └────┘         └────┘
```

We can compose them

```
  ┌───────────────────┐
a │ ┌╌╌╌╌┐ [b] ┌╌╌╌╌┐ │ [c]
──┼╌┤    ├╌╌>>╌┤    ├╌┼──>>
  │ └╌╌╌╌┘     └╌╌╌╌┘ │
  └───────────────────┘
```

where the outputs `bs :: [b]` of the first machine are is passed as inputs to the second machine, and processed all one by one.

> Notice that in fact everything works not only for lists, but for every foldable type.

Reviewer: ```suggestion
We use the `[b]` notation to denote multiple outputs, but we can in fact use any `Foldable` type.
```

In fact, we use [`singletons`](https://hackage.haskell.org/package/singletons) to promote `Topology` to the type level and use it to keep track of the available machine transitions at the type level.

We use a `Topology vertex` kind, and we construct types of that kind using the `Topology :: [(vertex, [vertex])] -> Topology vertex` type constructor. In such a use, notice that also `vertex` is a kind.

#### Defining a `Topology`

To define a `Topology`, we first need to define the type of its vertices.

> This is usually done by defining an enumeration of the vertices, like

Reviewer: ```suggestion
A straightforward approach is to create a sum type of vertices:
```

Author: I also thought initially to say `sum type`, but also `Either Int String` is a sum type and that probably won't work here because then it won't satisfy the `Demote vertex ~ vertex` constraint we have in the `Basic` constructor for `StateMachine`.

In fact, I need to add some more information about those constraints and what they imply (basically they imply that the type must be finite)
